openwisp_notifications/base/models.py

```diff
@@ -178,26 +180,71 @@ def redirect_view_url(self):
         )
 
 
-class AbstractNotificationUser(TimeStampedEditableModel):
+class AbstractNotificationSetting(UUIDModel):
     _RECEIVE_HELP = (
-        'note: non-superadmin users receive '
+        'Note: Non-superadmin users receive '
         'notifications only for organizations '
         'of which they are member of.'
     )
-    user = models.OneToOneField(settings.AUTH_USER_MODEL, on_delete=models.CASCADE)
-    receive = models.BooleanField(
-        _('receive notifications'), default=True, help_text=_(_RECEIVE_HELP)
+    user = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE)
+    type = models.CharField(
+        max_length=30,
+        null=True,
+        choices=NOTIFICATION_CHOICES,
+        verbose_name='Notification Type',
+    )
+    organization = models.ForeignKey(
+        get_model_name('openwisp_users', 'Organization'),
+        on_delete=models.CASCADE,
+        blank=True,
+        null=True,
+    )
+    web = models.BooleanField(
+        _('web notifications'), null=True, blank=True, help_text=_(_RECEIVE_HELP)
     )
     email = models.BooleanField(
-        _('email notifications'), default=True, help_text=_(_RECEIVE_HELP)
+        _('email notifications'), null=True, blank=True, help_text=_(_RECEIVE_HELP)
     )
 
     class Meta:
         abstract = True
+        constraints = [
+            UniqueConstraint(
+                fields=['organization', 'type', 'user'],
+                name='unique_notification_setting',
+            ),
+        ]
         verbose_name = _('user notification settings')
         verbose_name_plural = verbose_name
+        ordering = ['type', 'organization']
+
+    def __str__(self):
+        return '{type} {organization}'.format(
+            type=self.type,
+            organization=self.organization if self.organization else 'All',
+        )
 
     def save(self, *args, **kwargs):
-        if not self.receive:
-            self.email = self.receive
-        return super(AbstractNotificationUser, self).save(*args, **kwargs)
+        if not self.web_notification:
+            self.email = self.web_notification
+        return super().save(*args, **kwargs)
+
+    def full_clean(self, *args, **kwargs):
+        notification_type_config = get_notification_configuration(self.type)
```

Comment:
I'd create a property for this, eg:

`self.config`

Reply:
`self.config` can be misleading. I have used `self.notification_type_config` instead.